Feature 54050 password strength meter

LBruegger opened the pull request.

lucashabersaat commented on frontend/app/templates/components/account-form.hbs:
```diff
@@ -43,23 +40,23 @@
                     <div class="row">
-                      <Input class="form-control " @value={{this.changeset.cleartextPassword}} @name="cleartextPassword" @type="password"/>
+                      <Input class="form-control mb-10" @value={{mut this.changeset.cleartextPassword}} @name="cleartextPassword" @type="password" @key-up={{action "updatePasswordScore"}}/>
+                      {{ember-progress-bar progress=this.passwordScore options=(hash strokeWidth=3 duration=1)}}
```
Nice, I like the simple look. 
What about colors? Possible?
A subtle red, orange and green would be cool I think. But not sure if worth the effort...

LBruegger replied:
Possible. But you need to reload the ProgressBar everytime you want to change the Color.

lucashabersaat commented on frontend/app/templates/components/account-form.hbs:
```diff
@@ -43,23 +40,23 @@
                   <div class="form-group col-xs-12 col-sm-5 pull-right">
                     <div class="row">
                       <button class="btn btn-secondary btn-block" {{on "click" this.setRandomPassword}}>{{t "randomPassword"}}</button>
+                      <label>{{this.passwordLabel}}</label>
```
I would place the Label on one line right under the progress bar, instead under the Random Button.

LBruegger replied:
Sure.

lucashabersaat commented on frontend/app/styles/app.css:
```diff
@@ -40,3 +40,7 @@ li a {
   height: 69px;
 }
+
+.mb-10 {
```
Give it a clearer name? margin-bottom-10?

LBruegger replied:
We already have margin which is named like this. So i made it like them, to maintain constancy.

lucashabersaat commented on frontend/app/routes/accounts/new.js:
```diff
@@ -1,5 +1,12 @@
 import Route from '@ember/routing/route';
+import { getOwner } from '@ember/application';
 
 export default Route.extend({
+  get passwordStrength() {
+    return getOwner(this).lookup('service:password-strength');
+  },
 
+  beforeModel() {
+    return this.passwordStrength.load();
+  }
```
See above.

lucashabersaat commented on frontend/app/routes/accounts/edit.js:
```diff
@@ -1,6 +1,15 @@
+  get passwordStrength() {
+    return getOwner(this).lookup('service:password-strength');
+  },
+
```
Can you move this into _routes/accounts.js_ ? Not sure if that works.

lucashabersaat commented:
Another thing. If one sees this bar the first time, and doesn't change the password. It's not really clear, what this bar represents. Maybe add a Label StrengthMeter, or make the background grey to indicate more a progress bar. Something that shows, what this little box tells us.

psunix commented on frontend/app/components/account-form.js:
```diff
@@ -11,20 +11,38 @@ import { isPresent, isNone } from '@ember/utils';
+  @action
+  updatePasswordScore() {
+    if (isPresent(this.changeset.cleartextPassword)){
+      this.passwordStrength.strength(this.changeset.cleartextPassword).then(strength => {
```
can't we do this inside the service ?

psunix commented:
we really need to style this :smile: 
how about something like this ?

![image](https://user-images.githubusercontent.com/655627/82783130-2ef2cd00-9e5e-11ea-939e-9655c0ea22aa.png)

LBruegger commented:
I the last commit i fixed everything regarding the password strength meter. The logic for the password check is now available in the service. Also i reworked everything. Now it looks a lot better than my previous version.

But there is still one issue missing. I just didn't know how to open the bootstrap popover to check if the values are appropriate for each step of password strength.

psunix commented on frontend/app/components/account-form.js:
```diff
@@ -11,17 +11,53 @@ import { isPresent } from "@ember/utils";
+  }
+
+  popoverTemplate() {
+    return '<div class="popover"><div class="arrow"></div>'+
```
:nauseated_face: so much template code inside a component.

lucashabersaat commented on frontend/app/styles/app.css:
```diff
@@ -66,3 +74,30 @@ li a {
+.progress-bar-100 {
+  width: 100% !important;
+  background-color: #4dd100 !important;
+}
```
Not very happy with this, but it's safer instead of setting style inline. Wanted to put it in its own stylesheet, but didn't know how to in Ember. Hope that's ok for now.

lucashabersaat commented on frontend/app/components/password-strength-meter.js:
```diff
@@ -1,3 +1,41 @@
+    this.passwordStrength.load()
+  }
+
+  didUpdateAttrs() {
```
This is a hook, so must keep this name.
